shared/src/main/scala/scodec/InlineFoldR.scala

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2013, Scodec
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software without
+ *    specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package scodec
+
+import scala.compiletime.*
+
+object InlineFoldR:
+  trait Step[Acc[_ <: Tuple]]:
+    inline def apply[Elem, T <: Tuple](acc: Acc[T]): Acc[Elem *: T]
+
+  /** Inline version of foldRight that works on tuples */
+  inline def fold[Acc[_ <: Tuple], T <: Tuple](
+      acc: Acc[EmptyTuple],
+      f: Step[Acc]
+  ): Acc[T] =
+    inline erasedValue[T] match
+      case _: EmptyTuple => acc.asInstanceOf[Acc[T]]
+      case _: (h1 *: h2 *: h3 *: h4 *: h5 *: h6 *: h7 *: h8 *: h9 *: h10 *: h11 *: h12 *: h13 *:
```

Comment:
Why these specific arities (16, 4, 1)?

Reply:
These arities allows to use case classes up to `27 * 16 = 432` fields:
(in worst case for `431` field: `(26 inlines) * (by 16) + (3 inlines) * (by 4) + (3 inlines) * (by 1)`)

The question is what limit we really want?

Reply:
Ah I see, clever! Any way to match on `TupleXXL` and support an arbitrarily large tuple?

Reply:
I don't see how `TupleXXL` can help.
The problem is that we need to do some inline logic for each field of a tuple. When we need to traverse ordinary lists we use recursion or built-in methods.
In our case for inlining stuff there is only method `summonAll` which can summon something for each tuple field and which works for arbitrary tuple size. 
Hypothetically, we could try to invent some tricky typeclass to summon and `inline given` instance that performs needed logic.
But looks like unrolling loops is simpler and more predictable approach.